server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java

```diff
@@ -1050,4 +1006,91 @@ private void cleanUpCompactors() {
+    Set<ExternalCompactionId> idsInMetadata = readExternalCompactionIds();
+
+    var idsToRemove = Sets.difference(idsSnapshot, idsInMetadata);
+
+    // remove ids that are in the running set but not in the metadata table
+    idsToRemove.forEach(this::recordCompletion);
+
+    if (idsToRemove.size() > 0) {
+      LOG.debug("Removed stale entries from RUNNING_CACHE : {}", idsToRemove);
+    }
+
+    // Get the set of groups being referenced in the current configuration
+    // Needs Dan's changes for this
+    Set<CompactorGroupId> groupsInConfiguration = new HashSet<>();
+
+    // Compaction jobs are created in the TabletGroupWatcher and added to the Coordinator
+    // via the addJobs method which adds the job to the CompactionJobQueues object.
+    Set<CompactorGroupId> groupsWithJobs = jobQueues.getQueueIds();
+
+    Set<CompactorGroupId> jobGroupsNotInConfiguration =
+        Sets.difference(groupsWithJobs, groupsInConfiguration);
+
+    if (jobGroupsNotInConfiguration != null && !jobGroupsNotInConfiguration.isEmpty()) {
+      RUNNING_CACHE.values().forEach(rc -> {
+        if (jobGroupsNotInConfiguration.contains(CompactorGroupId.of(rc.getGroupName()))) {
+          LOG.warn(
+              "External compaction {} running in group {} on compactor {},"
+                  + " but group not found in current configuration. Failing compaction...",
+              rc.getJob().getExternalCompactionId(), rc.getGroupName(), rc.getCompactorAddress());
+          cancelCompactionOnCompactor(rc.getCompactorAddress(),
+              rc.getJob().getExternalCompactionId());
+        }
+      });
+
+      // Remove groups not in configuration from TIME_COMPACTOR_LAST_CHECKED
+      LOG.debug("No longer tracking compactor check-in times for groups: {}",
+          jobGroupsNotInConfiguration);
+      jobGroupsNotInConfiguration.forEach(TIME_COMPACTOR_LAST_CHECKED::remove);
+    }
+
+    final Set<CompactorGroupId> runningCompactorGroups = new HashSet<>();
+    getRunningCompactors().keySet()
+        .forEach(group -> runningCompactorGroups.add(CompactorGroupId.of(group)));
+
+    Set<CompactorGroupId> groupsWithNoCompactors =
+        Sets.difference(groupsInConfiguration, runningCompactorGroups);
+    if (groupsWithNoCompactors != null && !groupsWithNoCompactors.isEmpty()) {
+      LOG.warn("The following groups have no running compactors: {}", groupsWithNoCompactors);
+    }
+
+    Set<CompactorGroupId> compactorsWithNoGroups =
+        Sets.difference(runningCompactorGroups, groupsInConfiguration);
+    if (compactorsWithNoGroups != null && !compactorsWithNoGroups.isEmpty()) {
+      LOG.warn(
+          "The following groups have running compactors, but are not in the current configuration: {}",
+          compactorsWithNoGroups);
+    }
+
+    long now = System.currentTimeMillis();
+    Map<String,Set<HostAndPort>> idleCompactors = getIdleCompactors();
+    TIME_COMPACTOR_LAST_CHECKED.forEach((groupName, lastCheckTime) -> {
```

Comment:
Wonder if this should loop over idleCompactors at the top level instead of TIME_COMPACTOR_LAST_CHECKED.  Thinking there could be things coming back from idle compactors that are not in TIME_COMPACTOR_LAST_CHECKED.

```java
idleCompactors.forEach((group, compactors)->{
  long lastCeckTime = TIME_COMPACTOR_LAST_CHECKED.getOrDefault(group, 0);
  if ((now - lastCheckTime) > getMissingCompactorWarningTime()
          && jobQueues.getQueuedJobs(groupName) > 0){
     log.warn(...);
  }

});

```

Reply:
Actually we may want to loop over groupsInConfiguration at the top level.  This would find any configured group that has queued work and running compactors but no one is asking for work.

```
for(var group : groupsInConfiguration){
  long lastCeckTime = TIME_COMPACTOR_LAST_CHECKED.getOrDefault(group, 0);
  if ((now - lastCheckTime) > getMissingCompactorWarningTime()
          && jobQueues.getQueuedJobs(groupName) > 0
          && idleCompactors.contains(group)){
     log.warn(...);
  }
}

```

Reply:
Given an earlier suggestion that you made:
```
TIME_COMPACTOR_LAST_CHECKED.keySet().retainAll(groupsInConfiguration);
```

`TIME_COMPACTOR_LAST_CHECKED` only contains `groupsInConfiguration` (unless something new was added in between these two statements.

Reply:
> TIME_COMPACTOR_LAST_CHECKED only contains groupsInConfiguration (unless something new was added in between these two statements.

It could be a subset.    For example if 

 * groupsInConfiguration = [G1,G2]
 * TIME_COMPACTOR_LAST_CHECKED.keySet() = [G1,G3]

After we do retainAll, then TIME_COMPACTOR_LAST_CHECKED.keySet() will be [G1]. So it contains things we have seen, but not everything that is currently configured.